homeassistant/components/crownstone/entry_manager.py

```diff
@@ -0,0 +1,190 @@
+        asyncio.create_task(self.async_process_events(self.sse))
+        setup_sse_listeners(self)
+
+        # Set up a Crownstone USB only if path exists
+        if self.config_entry.data[CONF_USB_PATH] is not None:
+            await self.async_setup_usb()
+
+        # Save the sphere where the USB is located
+        # Makes HA aware of the Crownstone environment HA is placed in, a user can have multiple
+        self.usb_sphere_id = self.config_entry.data[CONF_USB_SPHERE]
+
+        self.hass.data.setdefault(DOMAIN, {})[self.config_entry.entry_id] = self
+        self.hass.config_entries.async_setup_platforms(self.config_entry, PLATFORMS)
+
+        # HA specific listeners
+        self.config_entry.async_on_unload(
+            self.config_entry.add_update_listener(_async_update_listener)
+        )
+        self.hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, self.on_shutdown)
+
+        return True
+
+    async def async_process_events(self, sse_client: CrownstoneSSEAsync) -> None:
+        """Asynchronous iteration of Crownstone SSE events."""
+        async with sse_client as client:
+            async for event in client:
+                if event is not None:
+                    # Make SSE updates, like ability change, available to the user
+                    self.hass.bus.async_fire(f"{DOMAIN}_{event.type}", event.data)
+
+    async def async_setup_usb(self) -> None:
+        """Attempt setup of a Crownstone usb dongle."""
+        # Trace by-id symlink back to the serial port
+        serial_port = await self.hass.async_add_executor_job(
+            get_port, self.config_entry.data[CONF_USB_PATH]
+        )
+        if serial_port is None:
+            return
+
+        self.uart = CrownstoneUart()
+        # UartException is raised when serial controller fails to open
+        try:
+            await self.uart.initialize_usb(serial_port)
+        except UartException:
+            delattr(self, ATTR_UART)
+            # Set entry data for usb to null
+            updated_entry_data = self.config_entry.data.copy()
+            updated_entry_data[CONF_USB_PATH] = None
+            updated_entry_data[CONF_USB_SPHERE] = None
+            # Ensure that the user can configure an USB again from options
+            self.hass.config_entries.async_update_entry(
+                self.config_entry, data=updated_entry_data, options={}
+            )
+            # Show notification to ensure the user knows the cloud is now used
+            persistent_notification.async_create(
+                self.hass,
+                f"Setup of Crownstone USB dongle was unsuccessful on port {serial_port}.\n \
+                Crownstone Cloud will be used to switch Crownstones.\n \
+                Please check if your port is correct and set up the USB again from integration options.",
+                "Crownstone",
+                "crownstone_usb_dongle_setup",
+            )
+            return
```

Comment:
Since you are checking that it worked when the config flow is setup, does it make sense to have them disable it in the config flow in case they accidentally knock it out of the device or their cat/dog/kid/so unplugs it to plug-in their other usb device?

Reply:
do you mean i should not reset the options here, but let the user manually do this? It would be an extra step to document. I have personally not tested your described case since i don't have a similar USB serial device currently. The window where the usb is unplugged and an other one is plugged in does seem really small though

Reply:
I'm thinking they have it up and running for a bit, and then the usb gets unplugged or knocked enough that it looses contact.  They probably expect to plug it back in and restart and be back up and running instead of getting their settings wiped without them making the change.

Reply:
Unplugging the USB while set up and running simply makes the UartManager thread wait for the USB to be plugged back in, and when you do, it automatically connects again, to the provided port. (without restart).

If the USB gets unplugged and HA is restarted with it unplugged, the config is still not removed, as `get_port()` in `async_setup_usb()` would return None, and it returns from the function, making the devices use cloud instead of USB. In that case the user would have to plug the USB back in and click on reload to make use of the USB again. (all of this was tested)

When using a manual port however, for example `/dev/test1234` for docker e.g. And it fails, it's mostly likely due to a typo in which case you would want to enter it again

Reply:
Is there a reason to allow manual entry?  It seems like pyserial is supported on all platforms that Home Assistant runs on so there shouldn't ever be a case were its not found?

Reply:
We had [this](https://github.com/crownstone/crownstone-home-assistant/issues/5) issue on current HACS plugin where a user created a custom udev rule for the dongle in a docker instance, which pyserial was unable to find

Reply:
Unless udev was improperly configured, it seems like it’s a bug in pyserial that it can’t find the device.  Was an issue opened for this?

Reply:
Will have to dive into this some more, i personally have not tested with docker and udev rule specifically, only python venv instance on Linux